libcudacxx/include/cuda/__memory_resource/cuda_memory_pool.h

```diff
@@ -0,0 +1,310 @@
+    __pool_properties.allocType     = ::cudaMemAllocationTypePinned;
+    __pool_properties.handleTypes   = ::cudaMemAllocationHandleType(__allocation_handle_type);
+    __pool_properties.location.type = ::cudaMemLocationTypeDevice;
+    __pool_properties.location.id   = __device_id;
+    ::cudaMemPool_t __cuda_pool_handle{};
+    _CCCL_TRY_CUDA_API(::cudaMemPoolCreate, "Failed to call cudaMemPoolCreate", &__cuda_pool_handle, &__pool_properties);
+
+    // CUDA drivers before 11.5 have known incompatibilities with the async allocator.
+    // We'll disable `cudaMemPoolReuseAllowOpportunistic` if cuda driver < 11.5.
+    // See https://github.com/NVIDIA/spark-rapids/issues/4710.
+    int __driver_version = 0;
+    _CCCL_TRY_CUDA_API(::cudaDriverGetVersion, "Failed to call cudaDriverGetVersion", &__driver_version);
+
+    constexpr int __min_async_version = 11050;
+    if (__driver_version < __min_async_version)
+    {
+      int __disable_reuse = 0;
+      _CCCL_TRY_CUDA_API(
+        ::cudaMemPoolSetAttribute,
+        "Failed to call cudaMemPoolGetAttribute",
+        __cuda_pool_handle,
+        ::cudaMemPoolReuseAllowOpportunistic,
+        &__disable_reuse);
+    }
+
+    _CCCL_TRY_CUDA_API(
+      ::cudaMemPoolSetAttribute,
+      "Failed to call cudaMemPoolReuseAllowOpportunistic",
+      __cuda_pool_handle,
+      ::cudaMemPoolAttrReleaseThreshold,
+      &__release_threshold);
+
+    // allocate the requested initial size to pprime the pool
+    if (__initial_pool_size != 0)
+    {
+      void* __ptr{nullptr};
+      _CCCL_TRY_CUDA_API(
+        ::cudaMallocAsync,
+        "cuda_memory_pool failed allocate the initial pool size",
+        &__ptr,
+        __initial_pool_size,
+        ::cudaStream_t{0});
+
+      _CCCL_ASSERT_CUDA_API(
+        ::cudaFreeAsync, "cuda_memory_pool failed to free the initial pool allocation", __ptr, ::cudaStream_t{0});
+    }
+    return __cuda_pool_handle;
+  }
+
+public:
+  /**
+   * @brief Constructs a cuda_memory_pool with the optionally specified initial pool size
+   * and release threshold.
+   *
+   * If the pool size grows beyond the release threshold, unused memory held by the pool will be
+   * released at the next synchronization event.
+   *
+   * @throws ::cuda::cuda_error if the CUDA version does not support `cudaMallocAsync`
+   *
+   * @param initial_pool_size Optional initial size in bytes of the pool. If no value is provided,
+   * initial pool size is half of the available GPU memory.
+   * @param release_threshold Optional release threshold size in bytes of the pool. If no value is
+   * provided, the release threshold is set to the total amount of memory on the current device.
+   */
+  cuda_memory_pool(
+    const size_t initial_pool_size                             = 0,
+    const size_t release_threshold                             = 0,
+    const cudaMemAllocationHandleType __allocation_handle_type = cudaMemAllocationHandleType::cudaMemHandleTypeNone)
```

Comment:
We should ensure that the constructor takes the device explicitly

Reply:
I moved to an API that explicitly takes the device id as a required argument. 

I found that it is easy to incorrectly mix up the other arguments, so I encapsulated them into a helper struct.

@harrism does that look reasonable to you?

Reply:
This will indeed suppress clang-tidy's `bugprone-easily-swappable-parameters` warning, which is nice. However I often struggle with the struct solution to that problem as I don't see how this helps avoid mixing up the two integral arguments:

```
cuda::mr::cuda_memory_pool_properties props = {42, 20};
```

At least it requires the caller to look at the struct and think about it, though. And I think the alternative of having strong types for the initial size and release threshold is overkill.

Reply:
I believe this is something where designated initializers shine
```cpp
cuda::mr::cuda_memory_pool_properties props = { .initial_pool_size = 42, .release_threshold = 20};
```